server/src/main/scala/org/broadinstitute/dsp/workbench/welder/server/StorageLinksService.scala

```diff
@@ -35,10 +40,27 @@ class StorageLinksService(storageLinks: StorageLinksCache) extends Http4sDsl[IO]
   //note: first param in the modify is the thing to do, second param is the value to return
   def createStorageLink(storageLink: StorageLink): IO[StorageLink] =
     storageLinks.modify { links =>
+
       val toAdd = List(storageLink.localBaseDirectory.path -> storageLink, storageLink.localSafeModeBaseDirectory.path -> storageLink).toMap
+
+      initializeDirectories(storageLink)
+
       (links ++ toAdd, storageLink)
     }
 
+  //returns whether the directories exist at the end of execution
+  def initializeDirectories(storageLink: StorageLink): Boolean = {
+    val localSafeAbsolutePath = workingDirectory.resolve(storageLink.localSafeModeBaseDirectory.path.asPath)
+    val localEditAbsolutePath = workingDirectory.resolve(storageLink.localBaseDirectory.path.asPath)
+
+    val dirsToCreate: List[java.nio.file.Path] = List[java.nio.file.Path](localSafeAbsolutePath, localEditAbsolutePath)
+
+    dirsToCreate
+      .map(path => new java.io.File(path.toUri))
+      .map(dir => dir.exists() || dir.mkdir()) //mkdir returns true if the directory is created. if it exists it returns false.
+      .foldRight(true)(_ && _)
```

Comment:
making directory is a side effect. wrap the action where actually creating directory in an `IO`, and return `IO[Unit]` (doesn't look like you care whether the directory is created successfully or not at call site)